File: libmedia/src/main/java/com/violet/libmedia/util/GLUtils.kt
package com.violet.libmedia.util

import android.opengl.GLES30
import com.violet.libbasetools.util.KLog

object GLUtils {
    const val TAG = "GLUtils"

    fun loadShader(type: Int, shaderSource: String): Int {
        val array = IntArray(1)
        // 1.创建一个着色器
        val shader = GLES30.glCreateShader(type)
        if (shader != 0) {
            // 2.提供着色器源码
            GLES30.glShaderSource(shader, shaderSource)
            // 3.编译着色器
            GLES30.glCompileShader(shader)
            // 4.获取着色器是否编译成功
            GLES30.glGetShaderiv(shader, GLES30.GL_COMPILE_STATUS, array, 0)
            if (array[0] == 0) {
                // 5.如果编译失败，获取编译失败的日志信息并删除着色器
                GLES30.glGetShaderiv(shader, GLES30.GL_INFO_LOG_LENGTH, array, 0)
                if (array[0] != 0) {
                    val info = GLES30.glGetShaderInfoLog(shader)
                    KLog.d(TAG, info)
                }
                GLES30.glDeleteShader(shader)
            }
        }
        return shader
    }

    fun createProgram(vertexShaderSource: String, fragmentShaderSource: String): Int {
        var program = 0
        var error = 0
        val array = IntArray(1)
        do {
            // 1.加载顶点着色器
            val vertexShader = loadShader(GLES30.GL_VERTEX_SHADER, vertexShaderSource)
            if (vertexShader == 0) {
                KLog.d(TAG, "load vertex shader failure")
                break
            }
            // 2.加载片段着色器
            val fragmentShader = loadShader(GLES30.GL_FRAGMENT_SHADER, fragmentShaderSource)
            if (fragmentShader == 0) {
                KLog.d(TAG, "load fragment shader failure")
                break
            }
            // 3.创建program对象
            program = GLES30.glCreateProgram()
            if (program == 0) {
                KLog.d(TAG, "glCreateProgram error")
                break
            }
            // 4.将着色器与program对象链接
            GLES30.glAttachShader(program, vertexShader)
            error = GLES30.glGetError()
            if (error != GLES30.GL_NO_ERROR) {
                KLog.d(TAG, "attach vertex shader error")
                break
            }
            GLES30.glAttachShader(program, fragmentShader)
            error = GLES30.glGetError()
            if (error != GLES30.GL_NO_ERROR) {
                KLog.d(TAG, "attach fragment shader error")
                break
            }

            GLES30.glLinkProgram(program)
            GLES30.glGetProgramiv(program, GLES30.GL_LINK_STATUS, array, 0)

            GLES30.glDetachShader(program, vertexShader)
            GLES30.glDeleteShader(vertexShader)
            GLES30.glDetachShader(program, fragmentShader)
            GLES30.glDeleteShader(fragmentShader)
            if (array[0] != GLES30.GL_TRUE) {
                GLES30.glGetProgramiv(program, GLES30.GL_INFO_LOG_LENGTH, array, 0)
                if (array[0] != 0) {
                    val info = GLES30.glGetProgramInfoLog(program)
                    KLog.d(TAG, info)
                }
            }

        } while (false)
        return program
    }
}

File: libmedia/src/main/cpp/component/base/GLUtils.cpp
//
// Created by bronyna on 2023/2/12.
//

#include "GLUtils.h"
#include "LogUtil.h"
#include <cstdlib>
#include <cstring>
#include <GLES2/gl2ext.h>

GLuint GLUtils::LoadShader(GLenum shaderType, const char *pSource) {
    GLuint shader = 0;
    shader = glCreateShader(shaderType);
    if (shader) {
        glShaderSource(shader, 1, &pSource, NULL);
        glCompileShader(shader);
        GLint compiled = 0;
        glGetShaderiv(shader, GL_COMPILE_STATUS, &compiled);
        if (!compiled) {
            GLint infoLen = 0;
            glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &infoLen);
            if (infoLen) {
                char *buf = (char *) malloc((size_t) infoLen);
                if (buf) {
                    glGetShaderInfoLog(shader, GL_INFO_LOG_LENGTH, &infoLen, buf);
                    LOGCATE("GLUtils::LoadShader Could not compile shader %d:\n%s\n",
                            shaderType, buf);
                    free(buf);
                }
                glDeleteShader(shader);
                shader = 0;
            }
        }
    }
    return shader;
}

GLuint GLUtils::CreateProgram(const char *pVertexShaderSource, const char *pFragShaderSource,
                              GLuint &vertexShaderHandle, GLuint &fragShaderHandle) {
    GLuint program = 0;
    vertexShaderHandle = LoadShader(GL_VERTEX_SHADER, pVertexShaderSource);
    if (!vertexShaderHandle) return program;
    fragShaderHandle = LoadShader(GL_FRAGMENT_SHADER, pFragShaderSource);
    if (!fragShaderHandle) return program;
    program = glCreateProgram();
    if (program) {
        glAttachShader(program, vertexShaderHandle);
        CheckGLError("glAttachShader");
        glAttachShader(program, fragShaderHandle);
        CheckGLError("glAttachShader");

        glLinkProgram(program);
        GLint linkStatus = GL_FALSE;
        glGetProgramiv(program, GL_LINK_STATUS, &linkStatus);

        glDetachShader(program, vertexShaderHandle);
        glDeleteShader(vertexShaderHandle);
        vertexShaderHandle = 0;
        glDetachShader(program, fragShaderHandle);
        glDeleteShader(fragShaderHandle);
        fragShaderHandle = 0;

        if (linkStatus != GL_TRUE) {
            GLint bufLength = 0;
            glGetProgramiv(program, GL_INFO_LOG_LENGTH, &bufLength);
            if (bufLength) {
                char *buf = (char *) malloc((size_t) bufLength);
                if (buf) {
                    glGetProgramInfoLog(program, bufLength, NULL, buf);
                    LOGCATE("GLUtils::CreateProgram Could not link program:\n%s\n", buf);
                    free(buf);
                }
            }
        }
    }
    LOGCATE("GLUtils::CreateProgram program = %d", program);
    return program;
}

void GLUtils::DeleteProgram(GLuint &program) {
    LOGCATE("GLUtils::DeleteProgram");
    if (program) {
        glUseProgram(0);
        glDeleteProgram(program);
        program = 0;
    }
}

void GLUtils::CheckGLError(const char *pGLOperation) {
    for (GLint error = glGetError(); error; error = glGetError()) {
        LOGCATE("GLUtils::CheckGLError GL Operation %s() glError (0x%x)\n", pGLOperation, error);
    }
}

GLuint GLUtils::CreateProgram(const char *pVertexShaderSource, const char *pFragShaderSource) {
    GLuint vertexShaderHandle, fragShaderHandle;
    return CreateProgram(pVertexShaderSource, pFragShaderSource, vertexShaderHandle,
                         fragShaderHandle);
}